plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,236 @@
+      return options;
+    }
+
+    const entityRefs = Array.isArray(options.recipients.entityRef)
+      ? options.recipients.entityRef
+      : [options.recipients.entityRef];
+
+    const outbound: Promise<void>[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        if (compoundEntityRef.kind !== 'group') {
+          return;
+        }
+
+        let destination;
+        try {
+          destination = await this.getChannelFromGroup(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for group entity: ${error.message}`,
+          );
+          return;
+        }
+
+        if (!destination) {
+          this.logger.debug(
+            `No Slack channel found for group entity: ${entityRef}`,
+          );
+          return;
+        }
+
+        const payload = JSON.stringify({
+          ...options.payload,
+          destination: destination,
+        });
+
+        this.logger.debug(`Sending Slack channel notification: ${payload}`);
+        outbound.push(
+          sendSlackNotification(this.channelWebhookUrl, payload).catch(
+            error => {
+              this.logger.error(
+                `Failed to send Slack channel notification: ${error.message}`,
+              );
+            },
+          ),
+        );
+      }),
+    );
+
+    // this is a bit fire and forget, but I don't think we want to block further processing.
+    await Promise.all(outbound);
+
+    if (outbound.length > 0) {
+      // If we sent team notifications, we don't want to send individual ones after resolving the group
+      // This is how we prevent the other methods from sending individual notifications.
+      // Given the current logic, if a Group has no slack channel id annotation, we'll let it get expanded to individuals.
+      options.recipients.entityRef = [];
+    }
+    return options;
+  }
+
+  async postProcess(
+    notification: Notification,
+    options: NotificationSendOptions,
+  ): Promise<void> {
+    if (options.recipients.type === 'broadcast' || !notification.user) {
+      return;
+    }
+
+    const destination = await this.getEmailFromUser(notification.user);
+    if (!destination) {
+      this.logger.debug(`No email found for user entity: ${notification.user}`);
+      return;
+    }
+
+    const payload = JSON.stringify({
+      ...notification.payload,
+      destination: destination,
+    });
+
+    this.logger.debug(`Sending Slack DM notification: ${payload}`);
+
+    await sendSlackNotification(this.directWebhookUrl, payload);
+  }
+
+  async getEntity(entityRef: string): Promise<Entity> {
+    const { token } = await this.auth.getPluginRequestToken({
+      onBehalfOf: await this.auth.getOwnServiceCredentials(),
+      targetPluginId: 'catalog',
+    });
+
+    const entity = await this.catalogClient.getEntityByRef(entityRef, {
+      token,
```

Comment:
You could also add the fields you need from these entities to minimize the data transfer size

Reply:
Yeah, I considered that. I also was considering using the bulk get endpoint, but I wanted to make sure that if there are N recipients if 1 of them has an issue we still send N-1 notifications. Will see what I can do.

Reply:
You could also do something like:
```ts
const users = catalogClient.getEntities(
      {
        filter: [
          {
            kind: 'User',
            [`spec.profile.email`]: CATALOG_FILTER_EXISTS,
          },
        ],
        fields: [
          'kind',
          'metadata.namespace',
          'metadata.name',
          `spec.profile.email`,
        ],
      },
      { token },
);

await cacheService.set('user_emails', users.items.reduce((acc, user) => {
  return acc[stringifyEntityRef(user)] = user.spec.profile.email;
}, {}), {ttl: 60000});
```

To fetch all users with email and cache them to object where the key is user entity reference and value is the email. This would only need one catalog API call every once in a while, depending on the TTL of cache.

Reply:
Also if you have all the emails of all users, you can send the broadcast notifications to all users pretty easily. Maybe something to consider?

Reply:
That would be one very large cache entry in larger orgs though. Not sure it's a good fit for some cache types

Reply:
Using `getEntitiesByRefs` behind a caching (with TTL) `DataLoader` sounds reasonable to me. Then you get automatic rollup into batches without having to care about it in your code.

That endpoint is very fast and handles missing items fine, it just returns null for those entries.

Reply:
We have about 10,000 users in our catalog and a much smaller subset would actually receive notifications, so I'm not sure I'd go that route for now. I did make some updates just testing them and will update the PR.

Reply:
> That would be one very large cache entry in larger orgs though. Not sure it's a good fit for some cache types

True, depends on cache type. Redis for example works pretty well for small cache items so probably caching the users one-by-one would be better option than the whole array under one key.

Reply:
```ts
cacheService.set(`${stringifyEntityRef(user)}_email`, user.spec.profile.email, {ttl: 60000});
```

Reply:
this is a great overhead fetching from local DB. The original code is just one short line: 'getEntityByRef'.
The cache and all the logic you display in the comments should be in the catalog API implementations. it is up to the client implementation or server implementation to maintain pools and caches.

Reply:
Do we expect such a large amount of notifications and do we expect it to be to so many users? because if not then the cache is useless here. The size of an entity is small. small text object.

Reply:
Not in the server. That is not something we're looking to provide.


Same with the client. Why would a general purpose client have a cache layer built in? Sounds like a recipe for confusion and bugs. You want to call the device and expect up to date data.

Reply:
The cache is nor about size, but about maintaining a sane request rate toward upstream services that hold data that very rarely changes.

Reply:
I updated the PR to add caching and limit the fields returned. The volume will really depend on implementations and use cases tbh.

Some of the use cases I've identified for us anyway:

* Announcements -- low volume, e.g. new features, PSAs
* Docs Feedback -- pretty low volume
* Pull Request dashboard -- could be much higher volume and many more users. I think caching could be useful here. We have our own internal plugin for this as our setup doesn't really allow us to use standard tooling or GHE's built in PRs view.

All in all, I'm not terribly concerned about the request volume tot he catalog as it seems to be able to handle quite a bit of traffic.